backends/ebpf/runtime/pcap_util.c

```diff
@@ -0,0 +1,237 @@
+#include "pcap_util.h"
+
+#define DLT_EN10MB 1
+
+/* Dynamically-allocated list of packets.
+ */
+struct pcap_list {
+    pcap_pkt **pkts;
+    uint32_t len;
+};
+
+/* An array of lists of packets */
+struct pcap_list_array {
+    pcap_list_t **lists;
+    uint16_t len;
+};
+
+pcap_list_t *append_packet(pcap_list_t *pkt_list, pcap_pkt *pkt) {
+    if (!pkt_list)
+        /* If the list is not allocated yet, create it */
+        pkt_list = allocate_pkt_list();
+    pkt_list->len++;
+    pkt_list->pkts = realloc(pkt_list->pkts, pkt_list->len * sizeof(pcap_pkt *));
+    if (pkt_list->pkts == NULL) {
+        fprintf(stderr, "Fatal: Failed to expand the"
+            "packet list with size %u !\n", pkt_list->len);
+        exit(EXIT_FAILURE);
+    }
+    pkt_list->pkts[pkt_list->len - 1] = pkt;
+    return pkt_list;
+}
+
+pcap_list_array_t *insert_list(pcap_list_array_t *pkt_array, pcap_list_t *pkt_list, uint16_t index) {
+    if (!pkt_array)
+        /* If the array is not allocated yet, create it */
+        pkt_array = allocate_pkt_list_array();
+    if (index >= pkt_array->len) {
+        pkt_array->lists = realloc(pkt_array->lists, (index+1) * sizeof(pcap_list_t *));
+        if (pkt_array->lists == NULL) {
+            fprintf(stderr, "Fatal: Failed to expand the "
+                "list array with size %u !\n", pkt_array->len);
+            exit(EXIT_FAILURE);
+        }
+        pkt_array->len = index+1;
+    }
+    pkt_array->lists[index] = pkt_list;
+    return pkt_array;
+}
+
+pcap_pkt *get_packet(pcap_list_t *list, uint32_t index) {
+    if (list->len < index) {
+        fprintf(stderr, "Index %u exceeds list size %u!\n", index, list->len);
+        return NULL;
+    }
+    return list->pkts[index];
+}
+
+pcap_list_t *get_list(pcap_list_array_t *array, uint16_t index) {
+    if (array->len < index) {
+        fprintf(stderr, "Index %u exceeds array size %u!\n", index, array->len);
+        return NULL;
+    }
+    return array->lists[index];
+}
+
+uint32_t get_pkt_list_length(pcap_list_t *pkt_list) {
+    return pkt_list->len;
+}
+
+uint16_t get_list_array_length(pcap_list_array_t *pkt_list_array) {
+    return pkt_list_array->len;
+}
+
+
+pcap_list_t *allocate_pkt_list() {
+    pcap_list_t *pkt_list = calloc(1, sizeof(pcap_list_t));
+    return pkt_list;
+}
+
+pcap_list_array_t *allocate_pkt_list_array() {
+    pcap_list_array_t *pkt_list_arr = calloc(1, sizeof(pcap_list_array_t));
+    return pkt_list_arr;
+}
+
+void delete_list(pcap_list_t *pkt_list) {
+    for(uint32_t i = 0; i< pkt_list->len; i++) {
+        free(pkt_list->pkts[i]->data);
+        /* Set the data pointer to NULL, to mitigate duplicate frees */
+        pkt_list->pkts[i]->data = NULL;
+        free(pkt_list->pkts[i]);
+    }
+    free(pkt_list->pkts);
+    free(pkt_list);
+}
+
+void delete_array(pcap_list_array_t *pkt_list_array) {
+    for(uint32_t i = 0; i< pkt_list_array->len; i++)
+        delete_list(pkt_list_array->lists[i]);
```

Comment:
This will crash if some lists were never allocated.
You must check for null lists either here or in delete_list.